### ISSUE TITLE
qs: show interesting strings first within each section

### ISSUE THREAD

**Issue by williballenthin** [mandiant/flare-floss#797]
via #761 and @r0ny123 

> Currently, QS ranks strings based on offsets within sections, can we use stringsifter or something like that to show the most relevant strings first within the sections?

**Comment by williballenthin** [comment]
this is a neat idea! keeping all the important data together should make it easier for humans to review.

i wonder if we can duplicate the important strings: show the group of important strings up top, and then continue to show all strings in linear order later on. for example:


```
32.exe                                                                                  #important         00028f5c              ┃┃┃┃┃┃┃┃┃┃┃┃
file size = %d bytes                                                                    #important         00029084              ┃┃┃┃┃┃
fseek(SEEK_SET) failed                                                                  #important         0002909c              ┃┃┃┃┃┃
---------------------------------------------------------------------------------------------------------------------------------┃┃┃┃┃┃
\\\\.\\a:                                                                                                  00028ec0              ┃┃┃┃┃┃
FILENAME                                                                                        #common    00028ec8              ┃┃┃┃┃┃
\\restore\\                                                                                                00028edc              ┃┃┃┃┃┃
Software\\Microsoft\\MSNetMng                                                                              00028ee8              ┃┃┃┃┃┃
Global\\DirectMarketing                                                                                    00028f08              ┃┃┃┃┃┃
Policy                                                                                          #common    00028f20              ┃┃┃┃┃┃
Version                                                                                           #curl    00028f28              ┃┃┃┃┃┃
Status                                                                                          #common    00028f34              ┃┃┃┃┃┃
Explorer.exe                                                                                    #common    00028f3c              ┃┃┃┃┃┃
32.exe                                                                                                     00028f5c              ┃┃┃┃┃┃
\\System32                                                                                      #common    00028f70              ┃┃┃┃┃┃
SYSTEMROOT                                                                                      #common    00028f7c              ┃┃┃┃┃┃
Software\\Microsoft\\Windows NT\\CurrentVersion\\Winlogon                                       #common    00028f90              ┃┃┃┃┃┃
System\\CurrentControlSet\\Services\\USBSTOR\\Enum                                                         00028fcc              ┃┃┃┃┃┃
ProductId                                                                                       #common    00029010              ┃┃┃┃┃┃
RegisteredOrganization                                                                          #common    0002901c              ┃┃┃┃┃┃
RegisteredOwner                                                                                 #common    00029034              ┃┃┃┃┃┃
Software\\Microsoft\\Windows NT\\CurrentVersion                                                 #common    00029044              ┃┃┃┃┃┃
file size = %d bytes                                                                                       00029084              ┃┃┃┃┃┃
fseek(SEEK_SET) failed                                                                                     0002909c              ┃┃┃┃┃┃
System\\CurrentControlSet\\Services\\PartMgr\\Enum                                                         000290b8              ┃┃┃┃┃┃
WS     TMP                                                                                                 000290e8              ┃┃┃┃┃┃
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛┃┃┃┃┃
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛┃┃┃┃
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛┃┃┃
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛┃┃
```

**Comment by vee1e** [comment]
this is a great idea! unfortunately stringsifter's model files are highly outdated and a pain to run on modern python and lightgbm versions. the maintainers haven't released any training files or the actual dataset used for it, but we do have this snippet here from the README

> This version of StringSifter was trained using Strings outputs from sampled malware binaries associated with the first [EMBER dataset](https://github.com/endgameinc/ember). Ordinal labels were generated using weak supervision procedures, and supervised learning is performed by [Gradient Boosted Decision Trees](https://github.com/microsoft/LightGBM) with a learning-to-rank objective function. See [Quick Links](https://github.com/mandiant/stringsifter#quick-links) for further technical details. Please note that neither labeled data nor training code is currently available, though we may reconsider this approach in future releases.

so we can do one of three things here to assist with this:

1. retrain and create a new model from scratch with the existing provided information and integrate it directly into QS with no external stringsifter dependency
2. harness the old stringsifter model files, build around it and maintain the project through updates
3. ask stringsifter maintainers to make the training files public (best option)

assuming (3) doesn't happen, (1) would require a lot of short term effort but would result in a cleaner QS codebase with less dependencies, and (2) is the quick and dirty fix that would require careful maintenance.

what do you think @williballenthin @mr-tz? i love the idea of stringsifter integration into FLOSS in general and want to do this justice, so i'd prefer option (1), and would like to work on it during the gsoc project if selected. do you think this would be better if just integrated like (2), and is the scope of this idea enough for gsoc's 12 week timeline?

**Comment by mr-tz** [comment]
IMO not a priority item. QS should alleviate most of the pain for strings analysis. Down the road QS output could likely be used to train a new string rating algorithm.

**Comment by williballenthin** [comment]
incidentally, for (1), i wonder if we could use the existing stringsifter as an oracle to train a new stringsifter. alternatively, perhaps we could use LLMs to score strings by "interestingness".

**Comment by vee1e** [comment]
but the EMBER dataset they used already comes pre-labeled, wouldn't using stringsifter as an oracle introduce more noise into our model?

**Comment by williballenthin** [comment]
i'm not sure if stringsifter used the "is malware" label on each file from EMBER (that seems smart) or if it there was a corpus of human labels applied to individual strings. 🤷🏻‍♂️
